### PR TITLE
feat(batch): add task level metrics delete 

### PR DESCRIPTION
**Pull request by ZENOTME**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

ref from #3832 
> * We don't need to care about deleting metrics in promethues, it will delete by promethues server after some preconfigured interval.
> * For deleting task level metrics, here is the changes:
>   a. Maintain one `BatchMetrics` for each `BatchTaskExecution`
>   b. When a batch task finished/aborted, move it to a deletion queue, which executes deletion of metrics after several minutes. It's important not to delete it immediately since promethues pull data periodically.

This pr add the task level metrics delete.

## Solution

- BatchTaskMetricsManager:
   BatchTaskMetricsManager is created in BatchEnvironment. When it created, it will start **deletor**. It's responsible to create(Register) task level metrics for every BatchExecution.
- BatchTaskMetrics
   Every time a BatchExecution is created, the BatchTaskMetricsManager will create BatchTaskMetrics(create_task_metrics) for it.
  BatchExecution can pass this BatchTaskMetrics to the executor, and the executor use this BatchTaskMetrics to record task level metrics. Before the executor record task level metrics, it will record the used label of the metrics in BatchTaskMetrics->record first. When the Execution is done, it will send all record it used to the delete queue.
- Deletor
  The deletor is spawned when the BatchTaskMetricsManager is created. It received the can-delete-record from the delete queue periodically(per 1min) and store them in delete_cache. It don't delete them immediately since in order to leave time for promethues to pull data. And next period before it received the can-delete-record, it will delete all record in delete_cache first.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#3832

### CODE REVIEW

File: src/batch/src/executor/monitor/stats.rs [hunk 1]
```diff
@@ -46,12 +44,189 @@ impl BatchMetrics {
...
+                                }
+                            }
+                            _ => {
+                                unimplemented!("Never reach here");
+                            }
+                        }
+                    }
+                }
+
+                // read from delete queue and push into delete_cache.
```
Comment (BowenXiao1999) [inline src/batch/src/executor/monitor/stats.rs:95]: Why design a cache here? What about directly delete when read from receiver?

File: src/batch/src/executor/monitor/stats.rs [hunk 2]
```diff
@@ -11,27 +11,25 @@
...
-impl BatchMetrics {
+impl BatchTaskMetricsManager {
     pub fn new(registry: Registry) -> Self {
-        let opts = histogram_opts!(
-            "batch_row_seq_scan_next_duration",
-            "Time spent deserializing into a row in cell based table.",
-            exponential_buckets(0.0001, 2.0, 20).unwrap() // max 52s
-        );
-        let row_seq_scan_next_duration = register_histogram_with_registry!(opts, registry).unwrap();
-
```
Comment (BowenXiao1999) [inline src/batch/src/executor/monitor/stats.rs:34]: row_seq_scan metrics is removed?

File: src/batch/src/executor/monitor/stats.rs [hunk 3]
```diff
@@ -46,12 +44,189 @@ impl BatchMetrics {
...
+            tokio::sync::mpsc::channel::<DeleteRecord>(1024);
+        let exchange_recv_row_number_clone = exchange_recv_row_number.clone();
+        tokio::spawn(async move {
+            let mut interval = tokio::time::interval(Duration::from_secs(60));
+            let mut delete_cache: Vec<DeleteRecord> = Vec::new();
+            let mut connect = true;
+            while connect {
+                // run every minute.
+                let _ = interval.tick().await;
+
```
Comment (BowenXiao1999) [inline src/batch/src/executor/monitor/stats.rs:64]: add some trace here?

File: src/batch/src/executor/monitor/stats.rs (hunk 1 above)
Comment (ZENOTME) [reply src/batch/src/executor/monitor/stats.rs:95]: When we read from receiver, we can't guarantee it have been pull by prometheus. (Think about a scene: A metrics just be create and the execution is done). So I put it in cache and then delete it next time to guarantee this metrics exist at least more than 1 min.

File: src/batch/src/executor/monitor/stats.rs (hunk 2 above)
Comment (ZENOTME) [reply src/batch/src/executor/monitor/stats.rs:34]: I move it in BatchTaskMetrics. Now we have two kinds of metrics:
- BatchMetrics
  exist in entire compute node lifecycle.
- BatchTaskMetrics
  exist in only a BatchExecution lifecycle. The recorded metrics will be deleted after execution done.

Comment (BowenXiao1999) [reply src/batch/src/executor/monitor/stats.rs:34]: But row_seq_scan should also be a per task metrics, isn't it? 🤔

File: src/batch/src/executor/monitor/stats.rs [hunk 4]
```diff
@@ -46,12 +44,189 @@ impl BatchMetrics {
...
+pub struct BatchTaskMetrics {
+    pub query_id: String,
+
+    /// Used to delete the label haved been recorded.
+    /// record definition:
+    /// HashMap { Key: metrics_name ,
+    ///           Value: the list of labels [ source_stage_id,
+    /// target_stage_id,source_task_id,target_task_id]         }
+    pub record: RwLock<HashMap<String, Vec<[u32; 4]>>>,
```
Comment (BowenXiao1999) [inline src/batch/src/executor/monitor/stats.rs:162]: can we make this private?

File: src/batch/src/executor/monitor/stats.rs (hunk 2 above)
Comment (ZENOTME) [reply src/batch/src/executor/monitor/stats.rs:34]: Seems it isn't. Because I look the use case of row_seq_scan and it didn't have any label like query_id or task_id.

File: src/batch/src/executor/monitor/stats.rs [hunk 5]
```diff
@@ -46,12 +44,189 @@ impl BatchMetrics {
...
+    pub exchange_recv_row_number: GenericCounterVec<AtomicU64>,
+    /// Used to send DeleteRecor to delete_queue.
+    delete_queue_sender: Sender<DeleteRecord>,
+}
+
+#[derive(Debug)]
+pub struct DeleteRecord {
+    query_id: String,
+    records: HashMap<String, Vec<[u32; 4]>>,
```
Comment (BowenXiao1999) [inline src/batch/src/executor/monitor/stats.rs:171]: Maybe make the Vector into a struct is better...

File: src/batch/src/executor/monitor/stats.rs [hunk 6]
```diff
@@ -46,12 +44,189 @@ impl BatchMetrics {
...
+}
+
+impl BatchTaskMetrics {
+    /// After we create a new label for a metric, we should record it first.
+    pub fn add_record(
```
Comment (BowenXiao1999) [inline src/batch/src/executor/monitor/stats.rs:176]: This API is a little strange, `BatchTaskMetrics` is for each task, but we get stage id and source/target etc..

What about we do not delete labels with such a detailed label, but in query level? e.g. if all tasks of query A is done in CN, delete all labels by query A?

File: src/batch/src/executor/monitor/stats.rs (hunk 5 above)
Comment (ZENOTME) [reply src/batch/src/executor/monitor/stats.rs:171]: Do you mean HashMap<String, Vec<Vec<u32>>>?

Comment (BowenXiao1999) [reply src/batch/src/executor/monitor/stats.rs:171]:
```rust
// Minimal unit for each delete
LabelDeleteUnit {
target_stage_id,
source_stage_id,
target_task_id,
source_task_id
}
```

File: src/batch/src/executor/monitor/stats.rs (hunk 6 above)
Comment (ZENOTME) [reply src/batch/src/executor/monitor/stats.rs:176]: But if we don't based on a detailed label, we can't statistics in a detailed level(such as task level). e.g If all task collect receive_row_number with label query_A, we only can see the sum of receive_row_number but receive_row_number of single task.

**Comment by liurenjie1024** [comment]
Good work, but I think we can improve this design.

For each `BatchTaskMetrics`, we just need to maintain a list of collector:
```
struct BatchTaskMetrics {
   collectors: Vec<Box<dyn Collector>>,
   /// Other fields
}
```

Then we can register/deregister collectors in registry: https://docs.rs/prometheus/latest/prometheus/struct.Registry.html#method.register

Also we can provide an task labels function in `BatchTaskContext` trait so that we can easily register metrics in executors. For example, we want to collect the actual map size in `HassAggExecutor`, we can do following:
```
let executor_labels = HashMap::with_capacity(5);
executor_labels.add_all(context.task_labels);
executor_labels.push("executor_id", executor_id);
let hash_map_size_opts = Opts::new("hash_agg_hash_map_size").const_labels(executor_labels);
let counter = register_counter(hash_map_size_opts);
```

**Comment by ZENOTME** [comment]
> Good work, but I think we can improve this design.
> 
> For each `BatchTaskMetrics`, we just need to maintain a list of collector:
> 
> ```
> struct BatchTaskMetrics {
>    collectors: Vec<Box<dyn Collector>>,
>    /// Other fields
> }
> ```
> 
> Then we can register/deregister collectors in registry: https://docs.rs/prometheus/latest/prometheus/struct.Registry.html#method.register
> 
> Also we can provide an task labels function in `BatchTaskContext` trait so that we can easily register metrics in executors. For example, we want to collect the actual map size in `HassAggExecutor`, we can do following:
> 
> ```
> let executor_labels = HashMap::with_capacity(5);
> executor_labels.add_all(context.task_labels);
> executor_labels.push("executor_id", executor_id);
> let hash_map_size_opts = Opts::new("hash_agg_hash_map_size").const_labels(executor_labels);
> let counter = register_counter(hash_map_size_opts);
> ```

This seems would cause  [AlreadyReg error](https://docs.rs/prometheus/latest/prometheus/struct.Registry.html#method.register:~:text=If%20the%20provided%20Collector%20is%20equal%20to%20a%20Collector%20already%20registered%20(which%20includes%20the%20case%20of%20re%2Dregistering%20the%20same%20Collector)%2C%20the%20AlreadyReg%20error%20returns.).
e.g. A compute node create 2 task and both of them register the "hash_agg_hash_map_size".

To solve this:
We defer the register job to deletor. When execution is done, BatchTaskMetrics will send the collector to deletor. the deletor get the collector and responsible to register them and unregister them. But this solution will cause the latency because every collector need to 'queue'.

**Comment by liurenjie1024** [comment]
I did some small example and it seems that it will not report error if labels are different.
```rust
#[cfg(test)]
mod tests {
    use prometheus::{labels, opts, register_counter_with_registry, Opts, Registry};

    #[test]
    fn register() {
        let register = Registry::new();

        let collector1 = {
            let opts = opts!(
                "counter",
                "help",
                labels! {
                  "query_id" => "a",
                  "stage_id" => "1"
                }
            );
            register_counter_with_registry!(opts, register).unwrap()
        };

        let collector2 = {
            let opts = opts!(
                "counter",
                "help",
                labels! {
                  "query_id" => "b",
                  "stage_id" => "1"
                }
            );

            register_counter_with_registry!(opts, register).unwrap()
        };

        register.unregister(Box::new(collector1.clone())).unwrap();
        register.unregister(Box::new(collector2.clone())).unwrap();
    }
}
```

File: src/batch/src/executor/monitor/stats.rs [hunk 7]
```diff
@@ -11,16 +14,140 @@
...
+}
+
+#[derive(Clone)]
+pub struct BatchTaskMetrics {
+    sender: Option<Sender<Box<dyn Collector>>>,
+    pub exchange_recv_row_number: GenericCounterVec<AtomicU64>,
+}
+
+impl BatchTaskMetrics {
+    pub fn new(registry: Registry, id: TaskId, sender: Option<Sender<Box<dyn Collector>>>) -> Self {
```
Comment (liurenjie1024) [inline src/batch/src/executor/monitor/stats.rs:105]: Why the `sender` is an `Option`?

File: src/batch/src/executor/monitor/stats.rs [hunk 8]
```diff
@@ -11,16 +14,140 @@
...
+                .send(Box::new(self.exchange_recv_row_number.clone()))
+                .await
+                .is_err()
+            {
+                error!("Failed to send delete record to delete queue");
+            }
+        }
+    }
+
+    pub fn unused() -> Self {
```
Comment (liurenjie1024) [inline src/batch/src/executor/monitor/stats.rs:144]: 1. Please add some comments to clarify that it's used for tests.
2. Add `#[cfg(test)]` so that it will not be misused by production code.
3. `unused` seems wierd to me, how about `for_test`?

File: src/batch/src/executor/monitor/stats.rs [hunk 9]
```diff
@@ -11,16 +14,140 @@
...
+        Self::new(prometheus::Registry::new(), TaskId::default(), None)
+    }
+}
 
 pub struct BatchMetrics {
     pub row_seq_scan_next_duration: Histogram,
```
Comment (liurenjie1024) [inline src/batch/src/executor/monitor/stats.rs:150]: I think we should move it to executor, but as discussed offline, we can put it in another pr.

File: src/batch/src/executor/monitor/stats.rs (hunk 3 above)
Comment (liurenjie1024) [reply src/batch/src/executor/monitor/stats.rs:50]: +1 Please add some info log here so that we can know whether it's alive in debugging.

**Review by liurenjie1024 — COMMENTED** [review]
Generally LGTM, please fix comments.

File: src/batch/src/executor/monitor/stats.rs (hunk 8 above)
Comment (ZENOTME) [reply src/batch/src/executor/monitor/stats.rs:144]: I find this function used in #[tokio::test] https://github.com/singularity-data/risingwave/blob/c56c90a99e3846455c94d76089bbd6496680cc38/src/compute/tests/integration_tests.rs#L220
If I add #[cfg(test)], this is invisible in #[tokio::test]. I will get the error like this:
```
error[E0599]: no function or associated item named `for_test` found for struct `risingwave_batch::executor::BatchMetrics` in the current scope
   --> src/compute/tests/integration_tests.rs:278:32
    |
278 |         Arc::new(BatchMetrics::for_test()),
    |                                ^^^^^^^^ function or associated item not found in `risingwave_batch::executor::BatchMetrics`
```
And I can't find the conditions like #[cfg(test)] to make it visible in #[tokio::test].
Do you some way to solve this, otherwise we can just remove #[cfg(test)] and add the comment.

Comment (liurenjie1024) [reply src/batch/src/executor/monitor/stats.rs:144]: Oh, I forget this. Yes we can only add comments in this case.

**Review by BowenXiao1999 — APPROVED** [review]

**Review by liurenjie1024 — APPROVED** [review]
LGTM

File: src/batch/src/executor/monitor/stats.rs [hunk 10]
```diff
@@ -11,16 +14,142 @@
...
+        let (delete_queue_sender, mut delete_queue_receiver) =
+            tokio::sync::mpsc::channel::<Box<dyn Collector>>(1024);
+        let deletor_registry = registry.clone();
+        tokio::spawn(async move {
+            let mut interval = tokio::time::interval(Duration::from_secs(60));
+            let mut delete_cache: Vec<Box<dyn Collector>> = Vec::new();
+            let mut connect = true;
+            while connect {
+                // run every minute.
+                trace!("BatchTaskMetricsManager Deletor is running...");
```
Comment (BowenXiao1999) [inline src/batch/src/executor/monitor/stats.rs:49]: tracing::info is better

File: src/batch/src/executor/monitor/stats.rs [hunk 11]
```diff
@@ -11,16 +14,146 @@
...
     register_int_counter_vec_with_registry, Histogram, Registry,
 };
+use tokio::sync::mpsc::Sender;
+
+use crate::task::TaskId;
+
+// When execution is done, it need to call clear_record() in BatchTaskMetrics.
+// The clear_record() will send the Collector to delete_queue, if the queue is full, the execution
+// will be blocked so that user can't get the result immediately.
+const DELETE_QUEUE_SIZE: usize = 4096;
```
Comment (ZENOTME) [inline src/batch/src/executor/monitor/stats.rs:29]: I find that the execution will be blocked when the delete queue is full. And that will caused the execution time be longer(can't not pass e2e test). 
For now, I set it to 4096 and seems it can pass the e2e test. But we may need better solution later. @liurenjie1024 @BowenXiao1999

Comment (liurenjie1024) [reply src/batch/src/executor/monitor/stats.rs:29]: I think another solution is to `select` futures of timers and deletion queue so that it will not block task clean up.

Comment (BowenXiao1999) [reply src/batch/src/executor/monitor/stats.rs:29]: Can we spawn a task to write the channel to avoid this?

Comment (liurenjie1024) [reply src/batch/src/executor/monitor/stats.rs:29]: You can take this for example.

Comment (ZENOTME) [reply src/batch/src/executor/monitor/stats.rs:29]: > I think another solution is to select futures of timers and deletion queue so that it will not block task clean up.
It may can alleviate. But it can't avoid some case where the collector is too much?

> Can we spawn a task to write the channel to avoid this?
I'm afraid that this will cause too much task.

Comment (liurenjie1024) [reply src/batch/src/executor/monitor/stats.rs:29]: I think we can make this queue unbounded, anyway it should exist in memory, so ubouned queue is ok to me. But we need to add a metric to monitor the pending tasks.
